smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/providers/helpers/PausablePollingStream.java

```diff
@@ -0,0 +1,144 @@
+package io.smallrye.reactive.messaging.providers.helpers;
+
+import static io.smallrye.reactive.messaging.providers.i18n.ProviderLogging.log;
+
+import java.time.Duration;
+import java.util.Queue;
+import java.util.concurrent.Flow;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.Uni;
+import io.smallrye.mutiny.helpers.queues.Queues;
+import io.smallrye.mutiny.operators.multi.processors.UnicastProcessor;
+
+/**
+ * A polling stream that can be paused and resumed.
+ *
+ * @param <P> polled item type
+ * @param <T> emitted item type
+ */
+public class PausablePollingStream<P, T> {
+
+    private static final int STATE_NEW = 0; // no request yet -- we start polling on the first request
+    private static final int STATE_POLLING = 1;
+    private static final int STATE_PAUSED = 2;
+    private static final int STATE_CANCELLED = 3;
+
+    private final AtomicInteger state = new AtomicInteger(STATE_NEW);
+    private final Queue<T> queue;
+    private final ScheduledExecutorService pollerExecutor;
+    private final String channel;
+    private final int maxQueueSize;
+    private final Uni<P> pollUni;
+    private final UnicastProcessor<T> processor;
+    private final Multi<T> stream;
+    private final int halfMaxQueueSize;
+    private final boolean pauseResumeEnabled;
+
+    /**
+     * Create a new polling stream.
+     *
+     * @param channel the channel name
+     * @param pollUni the uni that polls the data
+     * @param emitFunction the function that emits the polled data
+     * @param pollerExecutor the executor that polls the data
+     * @param maxQueueSize the maximum queue size
+     * @param pauseResumeEnabled whether pause/resume is enabled
+     */
+    public PausablePollingStream(String channel, Uni<P> pollUni,
+            BiConsumer<P, Flow.Processor<T, T>> emitFunction,
+            ScheduledExecutorService pollerExecutor,
+            int maxQueueSize,
+            boolean pauseResumeEnabled) {
+        this.channel = channel;
+        this.maxQueueSize = maxQueueSize;
+        this.halfMaxQueueSize = maxQueueSize / 2;
+        this.pollerExecutor = pollerExecutor;
+        this.pauseResumeEnabled = pauseResumeEnabled;
+        this.queue = Queues.createSpscUnboundedArrayQueue(maxQueueSize);
```

Comment:
Is it the right choice here? Why unbounded? 

\CC  @jponge.

Reply:
Perhaps because it's pausable and you don't want to loose any data? (at the cost or risking a OOME)

Reply:
This class is a generalization of the `KafkaRecordStreamSubscription`. 

From experience on the Kafka class, I think the maximum queue size should be poll size * 3.
As @jponge said I wanted to be defensive and not lose any data. 
Further work can be focused on improving the testing on this class and maybe replacing the `KafkaRecordStreamSubscription`.